presto-jmx/src/main/java/com/facebook/presto/connector/jmx/JmxHistoryDumper.java

```diff
@@ -0,0 +1,106 @@
+
+public class JmxHistoryDumper
+        implements Runnable
+{
+    private static final Logger log = Logger.get(JmxHistoryDumper.class);
+
+    private final JmxHistoricalData jmxHistoricalData;
+    private final JmxMetadata jmxMetadata;
+    private final JmxRecordSetProvider jmxRecordSetProvider;
+    private final ScheduledExecutorService executor = newSingleThreadScheduledExecutor();
+    private final long period;
+    private long lastDumpTimestamp;
+
+    @Inject
+    public JmxHistoryDumper(
+            JmxHistoricalData jmxHistoricalData,
+            JmxMetadata jmxMetadata,
+            JmxRecordSetProvider jmxRecordSetProvider,
+            JmxConnectorConfig jmxConfig)
+    {
+        this.jmxHistoricalData = requireNonNull(jmxHistoricalData, "jmxStatsHolder is null");
+        this.jmxMetadata = requireNonNull(jmxMetadata, "jmxMetadata is null");
+        this.jmxRecordSetProvider = requireNonNull(jmxRecordSetProvider, "jmxRecordSetProvider is null");
+        this.period = requireNonNull(jmxConfig, "jmxConfig is null").getDumpPeriod().toMillis();
+    }
+
+    public void start()
+    {
+        long nowMillis = getNowMillis();
+        long initialDelay = nowMillis - roundToPeriod(nowMillis + period);
+        executor.scheduleAtFixedRate(this, initialDelay, period, MILLISECONDS);
+    }
+
+    private long roundToPeriod(long millis)
+    {
+        return ((millis + period / 2) / period) * period;
+    }
+
+    @Override
+    public void run()
+    {
+        // we are using rounded up timestamp, so that records from different nodes and different
+        // tables will have matching timestamps (for joining/grouping etc)
+        long dumpTimestamp = roundToPeriod(getNowMillis());
+
+        // if something has lagged and next dump has the same timestamp as this one, ignore this dump
+        if (dumpTimestamp == lastDumpTimestamp) {
+            return;
+        }
+        lastDumpTimestamp = dumpTimestamp;
+
+        for (String tableName : jmxHistoricalData.getTables()) {
+            try {
+            JmxTableHandle tableHandle = requireNonNull(
+                    jmxMetadata.getTableHandle(new SchemaTableName(JmxMetadata.HISTORY_SCHEMA_NAME, tableName)),
+                    format("tableHandle is null for table [%s]", tableName));
+
+                List<Object> row = jmxRecordSetProvider.getLiveRow(
+                        tableHandle,
+                        jmxRecordSetProvider.getColumnTypes(tableHandle.getColumnHandles()),
+                        dumpTimestamp);
+                jmxHistoricalData.addRow(tableName, row);
+            }
+            catch (Throwable ex) {
+                log.error(ex, "Error starting JmxHistoryDumper");
+            }
+        }
+    }
+
+    public void shutdown()
+    {
+        executor.shutdown();
+    }
+
+    private static long getNowMillis()
+    {
+        return MILLISECONDS.convert(System.nanoTime(), NANOSECONDS);
```

Comment:
This is being used to record the timestamp for each row. You can't use System.nanoTime() that way. It doesn't necessarily have any bearing to "wall clock" time, and can only be used to measure intervals (i.e., record a timestamp, do something, record another one, and compute the delta between the two).

From the javadocs:

>   This method can only be used to measure elapsed time and is
     not related to any other notion of system or wall-clock time.
    The value returned represents nanoseconds since some fixed but
     arbitrary <i>origin</i> time (perhaps in the future, so values
     may be negative).  The same origin is used by all invocations of
     this method in an instance of a Java virtual machine; other
     virtual machine instances are likely to use a different origin.

You're going to have to use System.currentTimeMillis() for this.